tcpspy/tcpspy.8
.\" This file is part of tcpspy, a TCP/IP connection monitor.
.\"
.\" Copyright (c) 2000, 2001, 2002 Tim J. Robbins. 
.\" All rights reserved.
.\" 
.\" Redistribution and use in source and binary forms, with or without
.\" modification, are permitted provided that the following conditions
.\" are met:
.\" 1. Redistributions of source code must retain the above copyright
.\"    notice, this list of conditions and the following disclaimer.
.\" 2. Redistributions in binary form must reproduce the above copyright
.\"    notice, this list of conditions and the following disclaimer in the
.\"    documentation and/or other materials provided with the distribution.
.\" 3. The name of the author may not be used to endorse or promote products
.\"    derived from this software without specific prior written permission.
.\" 
.\" THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
.\" INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
.\" AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
.\" THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
.\" EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
.\" PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
.\" OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
.\" WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
.\" OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
.\" ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
.\"
.\" $Id: tcpspy.8,v 1.39 2001/05/20 11:52:12 tim Stab $
.TH TCPSPY 8 "May 2001" TJR "tcpspy 1.7"
.SH NAME
tcpspy \- TCP/IP Connection Monitor
.SH SYNOPSIS
.B tcpspy
.RB [ "-dp" ]
.RB [ "-e"
.IR "rule" ]...
.RB [ "-f"
.IR "rulefile" ]...
.RB [ "-F"
.IR "facility" ]
.RB [ "-I"
.IR "interval" ]
.RB [ "-U"
.IR "user" ]
.RB [ "-G"
.IR "group" ]
.SH DESCRIPTION
.I tcpspy
logs information about selected incoming and outgoing TCP/IP connections to
syslog. The following information is logged: username, local address and port,
remote address, port, and optionally the filename of the executable. At
present, only the IPv4 protocol is supported.
.PP
.SS "Options"
.TP
.BI -e " 'rule'"
Log only connections matching the specified rule. Rule syntax is outlined
below. If this option is specified more than once, connections matching any
of the specified rules are logged. You should quote the rule, as shown above.
.TP
.BI -f " rulefile"
Read rules from
.IR rulefile .
Each rule is on a new line. The `#' character may be used to add comments;
everything from this character to the end of the line is ignored.

The
.B -e
and
.B -f
options may be used together.
.TP
.BI -F " facility"
Log to syslog facility
.I facility
instead of the compile-time default setting. See the
.BR syslog.conf (5)
manual page for a list of facilities.
.TP
.BI -I " interval"
Update the internal state every
.I interval
milliseconds, instead of the default of 1000 ms. Connections that last less
than
.I interval
milliseconds may be missed, so you should experiment to find a value small
enough that it catches most connections, but not so small that it causes
tcpspy to use too much CPU time.
.TP
.BI -U " user"
Switch to the specified user after startup.
.I user
may be a numeric user id or a user name from the system password file.
.TP
.BI -G " group"
Switch to the specified group after startup.
.I group
may be a numeric group id or a group name from the system group file.
If a username to switch to with the
.B -U
option is specified but
.B -G
is omitted, tcpspy will switch to that specified user's primary group.
.TP
.B -d
Debugging mode; if this option is specified, tcpspy will not detach from the
console after initialisation, and will log connections to standard output
instead of syslog.
.TP
.B -p
Log the filename of the executable that created/accepted the connection.
You may require superuser privileges to obtain this information for processes
you do not own (this is a kernel limitation).

This option can greatly increase the amount of CPU time required to
process each connection/disconnection.

.PP
.SS "Rule Syntax"
A rule may be specified with the
.B -e
option to log information about connections matching this rule, overriding
the default of logging all connections.
.PP
The following comparison operations are defined:
.TP
.BI user " uid"
True if the local user initiating or accepting the connection has the
.B effective
user id
.IR uid .
.TP
.BI user " \N'34'username\N'34'"
Same as above, but using a username instead of a user id.
.TP
.BI lport " port"
True if the local end of the connection has port number
.IR port .
.TP
.BI lport " [low] - [high]"
True if the local end of the connection has a port number
greater than or equal to
.I low
and less than or equal to
.IR high .
If the form
.I low-
is used, high is assumed to be 65535.
If the form
.I -high
is used, low is assumed to be 0. It is an error to omit both
.IR low " and " high .
.TP
.BI lport " \N'34'service\N'34'"
Same as above, but using a service name from
.I /etc/services
instead of a port number.
.TP
.B rport
Same as
.B lport
but compares the port number of the remote end of the connection.
.TP
.BI laddr " n.n.n.n[/m.m.m.m]"
Interpreted as a "net/mask" expression; true if "net" is equal to the bitwise
AND of the local address of the connection and "mask". If no mask is specified,
a default mask with all bits set (255.255.255.255) is used.
.TP
.B raddr
Same as
.B laddr
but compares the remote address.
.TP
.BI exe " \N'34'pattern\N'34'"
True if the full filename (including directory) of the executable that 
created/accepted the connection matches
.IR pattern ,
a
.BR glob (7)-style
wildcard pattern.

The pattern "" (an empty string) matches connections created/accepted by 
processes whose executable filename is unknown.

If the
.B -p
option is not specified, a warning message will be printed, and the result of
this comparison will always be true.
.PP
Expressions (including the  comparisons listed above) may be joined together
with the following logical operations:
.TP
.IB expr1 " or " expr2
True if either of
.I expr1
or
.I expr2
are true (logical OR).
.TP
.IB expr1 " and " expr2
True if both
.I expr1
and
.I expr2
are true (logical AND).
.TP
.BI not " expr"
True if
.I expr
is false (logical NOT).
.PP
Rules are evaluated from left to right. Whitespace (space, tab and newline)
characters are ignored between "words". Rules consisting of only whitespace
match no connections, but do not cause an error.
Parentheses, '(' and ')' may be placed around expressions to affect the order
of evaluation.
.PP
The Examples section contains some sample rules which further demonstrate how
they are constructed.

.SH "EXIT STATUS"
.TP
0
The daemon was successfully started
.TP
>0
An error occurred

.SH SIGNALS
.TP
.I TERM
Shut down at most
.I interval
milliseconds from now.
.TP
.I INT
(Debugging mode only) Handled identically to
.IR TERM .
.PP
All other signals retain their default behaviour, which is documented in
.BR signal (7).

.SH EXAMPLES
.TP
tcpspy -e 'user "joe" and rport "ssh"'
Log connections made by user "joe" for the service "ssh".
.TP
tcpspy -e 'not raddr 10.0.0.0/255.0.0.0 and rport 25 and (user "bob" or user "joe")'
Log connections made by users "bob" and "joe" to remote port 25 on machines
not on a fictional "intranet".
.TP
tcpspy -e 'exe "/usr/bin/irc"'
Log connections made by /usr/bin/irc (probably ircII).

.SH BUGS
Empty rule files cause
.B tcpspy
to log no connections instead of all connections.

.SH AUTHOR
Tim J. Robbins <tim@robbins.dropbear.id.au>

.SH SEE ALSO
.BR glob (7),
.BR proc (5),
.BR services (5),
.BR signal (7),
.BR syslog (3),
.BR syslog.conf (5)


tcpspy/tcpspy.c
/*
 * tcpspy.c - TCP/IP connection monitor.
 *
 * This file is part of tcpspy, a TCP/IP connection monitor.
 *
 * Copyright (c) 2000, 2001, 2002 Tim J. Robbins. 
 * All rights reserved.
 *
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions
 * are met:
 * 1. Redistributions of source code must retain the above copyright
 *    notice, this list of conditions and the following disclaimer.
 * 2. Redistributions in binary form must reproduce the above copyright
 *    notice, this list of conditions and the following disclaimer in the
 *    documentation and/or other materials provided with the distribution.
 * 3. The name of the author may not be used to endorse or promote products
 *    derived from this software without specific prior written permission.
 *
 * THIS SOFTWARE IS PROVIDED ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES,
 * INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY
 * AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.  IN NO EVENT SHALL
 * THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL,
 * EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO,
 * PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS;
 * OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY,
 * WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR
 * OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF
 * ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 *
 * $Id: tcpspy.c,v 1.39.1.3 2001/07/02 11:55:36 tim Stab $
 */

//superspy addition
#include <sqlite3.h>
//----------------

#include <arpa/inet.h>
#include <assert.h>
#include <ctype.h>
#include <dirent.h>
#include <errno.h>
#include <grp.h>
#include <limits.h>
#include <netdb.h>
#include <netinet/in.h>
#include <netinet/tcp.h>
#include <pwd.h>
#include <signal.h>
#include <stdarg.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#define SYSLOG_NAMES
#include <syslog.h>
#include <sys/ioctl.h>
#include <sys/socket.h>
#include <sys/time.h>
#include <sys/types.h>
#include <unistd.h>

#include "log.h"
#include "rcsid.h"
#include "rule.h"

RCSID("$Id: tcpspy.c,v 1.39.1.3 2001/07/02 11:55:36 tim Stab $");

/*
 * Defaults for compile-time settings. Descriptions of these are in
 * the Makefile.
 */
#ifndef CONNTABLE_BUCKETS
#define CONNTABLE_BUCKETS 5003
#endif
#ifndef FACILITY
#define FACILITY LOG_USER
#endif

int gotrule = 0;

static int stopped = 0, showprocs = 0, db = 0;
static int facility = FACILITY;

/*
 * This structure holds everything we need to know about a connection. We
 * use unsigned long instead of (for example) uid_t, ino_t to make hashing
 * easier.
 */
typedef struct conn {
	unsigned long lcl;
	unsigned long lclp;
	unsigned long rmt;
	unsigned long rmtp;
	unsigned long uid;
	unsigned long ino;

	char exe[PATH_MAX];
	
	struct conn *next;
} conn_t;

typedef struct conntable {
	conn_t *buckets[CONNTABLE_BUCKETS];
} conntable_t;

static int ct_init (conntable_t *ct);
static int ct_hash (conn_t *c);
static int ct_add (conntable_t *ct, conn_t *c);
static int ct_find (conntable_t *ct, conn_t *c);
static int ct_read (conntable_t *ct);
static int ct_free (conntable_t *ct);
static void huntinode (ino_t i, char *buf, size_t bufsize);
static void logconn (conn_t *c, const char *action);
static void logdb (conn_t *c, const char *action);
static void compare (conntable_t *old, conntable_t *new);
static void stopsig (int signal);

/*
 * stopsig ()
 *
 * Stop the program when certain signals are received. 
 */
static void stopsig (int signo)
{
	logmsg ("caught signal %d, shutting down", signo);
	stopped = 1;
}

/*
 * ct_init ()
 *
 * Initialise a connection table (hashtable).
 */
static int ct_init (conntable_t *ct)
{
	int i;
	
	assert (ct != NULL);

	for (i = 0; i < CONNTABLE_BUCKETS; i++)
		ct->buckets[i] = NULL;

	return 1;
}

/*
 * ct_hash ()
 *
 * Simple hash function for connections.
 */
static int ct_hash (conn_t *c)
{
	unsigned long h;

	assert (c != NULL);

	h = c->lcl ^ c->lclp ^ c->rmt ^ c->rmtp ^ c->uid ^ c->ino;

	return h % CONNTABLE_BUCKETS;
}

/*
 * ct_add ()
 *
 * Add a connection to the connection table.
 */
static int ct_add (conntable_t *ct, conn_t *c)
{
	conn_t *old, *newc;
	int bi;
	
	assert (ct != NULL);
	assert (c != NULL);

	newc = (conn_t *) malloc (sizeof (conn_t));
	if (newc == NULL) panic ("memory exhausted");	

	memcpy (newc, c, sizeof (conn_t));

	bi = ct_hash (c);
	old = ct->buckets[bi];
	ct->buckets[bi] = newc;
	ct->buckets[bi]->next = old;

	return 1;
}

/*
 * ct_find ()
 * 
 * Find a connection in a table, return nonzero if found, zero otherwise.
 */
static int ct_find (conntable_t *ct, conn_t *c)
{
	conn_t *bucket;
	
	assert (ct != NULL);
	assert (c != NULL);

	bucket = ct->buckets[ct_hash (c)];
	while (bucket != NULL) {
		if ((c->lcl == bucket->lcl) && (c->lclp == bucket->lclp) && 
			(c->rmt == bucket->rmt) && (c->rmtp == bucket->rmtp) &&
			(c->uid == bucket->uid) && (c->ino == bucket->ino)) {
			return 1;
		}
		bucket = bucket->next;
	}

	return 0;
}

/*
 * ct_read ()
 * 
 * Read /proc/net/tcp and add all connections to the table if connections
 * of that type are being watched.
 */
static int ct_read (conntable_t *ct)
{
	static FILE *fp = NULL;
	char buf[1024];
	conn_t c;
	
	assert (ct != NULL);

	if (fp == NULL) {
		fp = fopen ("/proc/net/tcp", "r");
		if (fp == NULL) panic ("/proc/net/tcp: %s", strerror (errno));
	}
	rewind (fp);

	if (fgets (buf, sizeof (buf), fp) == NULL)
		panic ("/proc/net/tcp: missing header");
	
	while (fgets (buf, sizeof (buf), fp) != NULL) {
		unsigned long st;

		if (sscanf (buf, "%*d: %lx:%lx %lx:%lx %lx %*x:%*x %*x:%*x %*x %lu %*d %lu", &c.lcl, &c.lclp, &c.rmt, &c.rmtp, &st, &c.uid, &c.ino) != 7) {
			logmsg ("/proc/net/tcp: warning: incomplete line");
			continue;
		}
		if ((c.ino == 0) || (st != TCP_ESTABLISHED)) continue;
		if (showprocs != 0)
			huntinode ((ino_t) c.ino, c.exe, sizeof (c.exe));

		if (ct_add (ct, &c) == 0)
			return 0;
	}

	return 1;
}

/*
 * ct_free ()
 *
 * Free a connection table.
 */
static int ct_free (conntable_t *ct)
{
	int i;

	assert (ct != NULL);

	for (i = 0; i < CONNTABLE_BUCKETS; i++) {
		conn_t *c0, *c1;

		c0 = ct->buckets[i];
		while (c0 != NULL) {
			c1 = c0->next;
			free (c0);
			c0 = c1;
		}
		ct->buckets[i] = NULL;
	}

	return 1;
}

/*
 * huntinode ()
 *
 * Find names processes using an inode and put them in a buffer.
 */
static void huntinode (ino_t i, char *buf, size_t bufsize)
{
	DIR *procdir;
	struct dirent *procent;
	
	assert (buf != NULL);
	*buf = '\0';

	if ((procdir = opendir ("/proc")) == NULL)
		panic ("/proc: %s", strerror (errno));
	while ((procent = readdir (procdir)) != NULL) {
		char fdbuf[PATH_MAX];
		DIR *fddir;
		struct dirent *fdent;

		/*
		 * No test needed for "." and ".." since they don't begin
		 * with digits.
		 */
		if (! isdigit (*procent->d_name))
			continue;
		
		snprintf (fdbuf, sizeof (fdbuf), "/proc/%s/fd", 
				procent->d_name);
		
		/*
		 * We're don't always run as root, we may get EPERM here,
		 * ignore it.
		 */
		if ((fddir = opendir (fdbuf)) == NULL)
			continue;

		while ((fdent = readdir (fddir)) != NULL) {
			int len;
			char lnkbuf[PATH_MAX], lnktgt[PATH_MAX];
			char exebuf[PATH_MAX], exetgt[PATH_MAX];
			ino_t this_ino;
			
			if (! isdigit (*fdent->d_name))
				continue;
			snprintf (lnkbuf, sizeof (lnkbuf), "%s/%s", fdbuf, 
					fdent->d_name);
			len = readlink (lnkbuf, lnktgt, sizeof (lnktgt) - 1);
			if (len < 0)
				continue;
			lnktgt[len] = '\0';
			if (sscanf (lnktgt, "socket:[%lu]", &this_ino) != 1)
				continue;
			if (this_ino != i)
				continue;

			snprintf (exebuf, sizeof (exebuf), "/proc/%s/exe", 
					procent->d_name);
			len = readlink (exebuf, exetgt, sizeof (exetgt) - 1);
			if (len < 0)
				continue;
			exetgt[len] = '\0';

			strncpy (buf, exetgt, bufsize);
			buf[bufsize - 1] = '\0';
		}

		closedir (fddir);
	}
	closedir (procdir);
}

/*
 * logconn ()
 *
 * Log a connection/disconnection.
 */
static void logconn (conn_t *c, const char *action)
{
	struct in_addr in;
	char laddr[16], raddr[16]; /* assume IPv4 nnn.nnn.nnn.nnn\0 */
	char *n;
	struct passwd *pw;
	char uidbuf[32];
	char *user;
	struct servent *se;
	char lport[64], rport[64];

	assert (c != NULL);
	assert (action != NULL);

	if ((gotrule != 0) && (rule_eval (c->uid, c->lcl, c->lclp, c->rmt, c->rmtp, (showprocs != 0) ? c->exe : NULL) == 0))
		return;

	in.s_addr = c->lcl;
	n = inet_ntoa (in);
	strncpy (laddr, n, sizeof (laddr));
	laddr[sizeof (laddr) - 1] = '\0';
	in.s_addr = c->rmt;
	n = inet_ntoa (in);
	strncpy (raddr, n, sizeof (raddr));
	raddr[sizeof (raddr) - 1] = '\0';

	snprintf (lport, sizeof (lport), "%lu", c->lclp);
	snprintf (rport, sizeof (rport), "%lu", c->rmtp);

	if ((se = getservbyport (htons (c->lclp), "tcp")) != NULL) {
		strncpy (lport, se->s_name, sizeof (lport));
		lport[sizeof (lport) - 1] = '\0';
	}
	if ((se = getservbyport (htons (c->rmtp), "tcp")) != NULL) {
		strncpy (rport, se->s_name, sizeof (rport));
		rport[sizeof (rport) - 1] = '\0';
	}

	if ((pw = getpwuid ((uid_t) c->uid)) == NULL) {
		snprintf (uidbuf, sizeof (uidbuf),
				"(uid %u)", (unsigned int)(uid_t)c->uid);
		user = uidbuf;
	} else {
		user = pw->pw_name;
	}

	if (showprocs != 0)
		logmsg ("%s: proc %.30s, user %s, local %s:%s, remote %s:%s",
			action, *c->exe != '\0' ? c->exe : "(unknown)", 
			user, laddr, lport, raddr, rport);
	else
		logmsg ("%s: user %s, local %s:%s, remote %s:%s",
				action, user, laddr, lport, raddr, rport);
}
/*
 * logdb ()
 *
 * Log a connection/disconnection to the database.
 */
static void logdb (conn_t *c, const char *action)
{

	struct in_addr in;
	char laddr[16], raddr[16]; /* assume IPv4 nnn.nnn.nnn.nnn\0 */
	char *n;
	struct passwd *pw;
	char uidbuf[32];
	char *user;
	struct servent *se;
	char lport[64], rport[64];


	//db variables
	int error=0;
	sqlite3 *DB;
	sqlite3_stmt *stmt;
	char lclstring[16];
	char lclpstring[64]; 
	char rmtstring[16];
	char rmtpstring[64]; 
	//db variables

	assert (c != NULL);
	assert (action != NULL);


	//open the database "super_spy.db"
	error = sqlite3_open("super_spy.db", &DB);
		if (error)
		{
                	printf("Can not open database, SQL error: ");
			printf(" %d\n", error);
		}

	if ((gotrule != 0) && (rule_eval (c->uid, c->lcl, c->lclp, c->rmt, c->rmtp, (showprocs != 0) ? c->exe : NULL) == 0))
		return;

	in.s_addr = c->lcl;
	n = inet_ntoa (in);
	strncpy (laddr, n, sizeof (laddr));
	laddr[sizeof (laddr) - 1] = '\0';
	in.s_addr = c->rmt;
	n = inet_ntoa (in);
	strncpy (raddr, n, sizeof (raddr));
	raddr[sizeof (raddr) - 1] = '\0';

	snprintf (lport, sizeof (lport), "%lu", c->lclp);
	snprintf (rport, sizeof (rport), "%lu", c->rmtp);

	sprintf(lclpstring, "%s", lport);//addition
	sprintf(rmtpstring, "%s", rport);//addition
	sprintf(lclstring, "%s", laddr);//addition
	sprintf(rmtstring, "%s", raddr);//addition


	// Create a prepared statement for db
	error = sqlite3_prepare_v2(DB, "insert into tcp values (?, ?, ?, ?, ?);", -1, &stmt, NULL);
		if (error != SQLITE_OK)
		{
			printf("error = sqlite3_prepare_v2: ");
			printf("%d\n", error);
		}
		//bind variables to columns in db
			 sqlite3_bind_text(stmt, 2, lclstring, 100, SQLITE_STATIC); 
			 sqlite3_bind_text(stmt, 3, lclpstring, 100, SQLITE_STATIC); 
			 sqlite3_bind_text(stmt, 4, rmtstring, 100, SQLITE_STATIC); 
			 sqlite3_bind_text(stmt, 5, rmtpstring, 100, SQLITE_STATIC); 


		// Execute the sql statement to fill super_spy.db
		error = sqlite3_step(stmt);
			if (error != SQLITE_DONE)
			{
				printf("error = sqlite3_step: ");
				printf("%d\n", error);				
			}
		//reset sql statement for next turn in loop
		sqlite3_reset(stmt);

		// Free the sql statement
		sqlite3_finalize(stmt);

		//Close db
		sqlite3_close(DB);
}

/*
 * compare ()
 *
 * Compare the `old' and `new' tables, logging any differences.
 */
static void compare (conntable_t *old, conntable_t *new)
{
	int i;
	
	assert (old != NULL);
	assert (new != NULL);
	
	for (i = 0; i < CONNTABLE_BUCKETS; i++) 
	{
		conn_t *bucket;

		bucket = old->buckets[i];
		while (bucket != NULL) 
		{
			if (ct_find (new, bucket) == 0)
				logconn (bucket, "disconnect");	
			bucket = bucket->next;
		}
	}

	for (i = 0; i < CONNTABLE_BUCKETS; i++) 
	{
		conn_t *bucket;

		bucket = new->buckets[i];
		while (bucket != NULL) 
		{
			if (ct_find (old, bucket) == 0)
			{
				logconn (bucket, "connect");
				if(db != 0)
				{
					logdb (bucket, "connect");
				}
			}
			bucket = bucket->next;
		}
	}
}

/*
 * getfacility ()
 *
 * Convert a facility name to a integer facility value for syslog().
 */
int getfacility (const char *s)
{
	int i;

	assert (s != NULL);

	for (i = 0; i < LOG_NFACILITIES - 2; i++) {
		assert (facilitynames[i].c_name != NULL);
		if (strcmp (facilitynames[i].c_name, s) == 0)
			return facilitynames[i].c_val;
	}

	return -1;
}

static void usage (void)
{
	fprintf (stderr, "usage: tcpspy [-dp] [-e rule]... [-f rulefile] [-I interval] "
			"[-U user]\n\t[-G group] [-F facility] [-I interval]\n");
	exit (EXIT_FAILURE);
}

int firstrule = 1;

int main (int argc, char *argv[])
{
	conntable_t old, new;
	unsigned long interval = 1000;
	int ch;
	uid_t dropuser = -1;
	gid_t dropgroup = -1, defgroup = -1;
	int debug = 0;

	log_set_syslog ();

	/*
	 * Parse our arguments.
	 */
	opterr = 0;
	while ((ch = getopt (argc, argv, "dpze:I:U:G:u:w:i:f:F:")) != -1) {
		switch (ch) {
			case 'd':
				debug = 1;
				log_set_stdout ();
				break;
			case 'p':
				showprocs = 1;
				break;
			case 'z':
				db = 1;
				break;
			case 'e':
				rule_parse (optarg);
				gotrule = 1;
				if (firstrule == 0) 
					/*
					 * Join multiple rules together with
					 * logical OR's.
					 */
					rule_gen_or ();
				firstrule = 0;
				/* Hide rule from `ps' */
				memset (optarg, '\0', strlen (optarg));
				break;
			case 'f':
			{
				FILE *fp;

				if ((fp = fopen (optarg, "r")) == NULL) {
					fprintf (stderr, "tcpspy: %s: %s\n", optarg, strerror (errno));
					exit (EXIT_FAILURE);
				}

				rule_parse_file (fp);

				fclose (fp);
			}
			break;
			case 'I':
				interval = atoi (optarg);
				if (interval == 0) {
					fprintf (stderr, "tcpspy: bad interval\n");
					exit (EXIT_FAILURE);
				}
				break;
			case 'U':
				{
				struct passwd *pw;
				
				if (isdigit (*optarg)) {
					dropuser = atoi (optarg);
					pw = getpwuid (atoi (optarg));
				} else {
					if ((pw = getpwnam (optarg)) == NULL) {
						fprintf (stderr, "tcpspy: user `%s' unknown\n", optarg);
						exit (EXIT_FAILURE);
					}
					dropuser = pw->pw_uid;
				}
				
				/*
				 * Use the gid from the password file entry if
				 * possible, as a default.
				 */
				if (pw != NULL)
					defgroup = pw->pw_gid;
				else
					defgroup = (gid_t) -1;
				
				}
				break;
			case 'G':
				if (isdigit (*optarg))
					dropgroup = atoi (optarg);
				else {
					struct group *gr;

					if ((gr = getgrnam (optarg)) == NULL) {
						fprintf (stderr, "tcpspy: group `%s' unknown\n", optarg);
						exit (EXIT_FAILURE);
					}
					dropgroup = gr->gr_gid;
				}	
				break;
			case 'F':
				if ((facility = getfacility (optarg)) < 0) {
					fprintf (stderr, "tcpspy: bad facility name `%s'\n", optarg);
					exit (EXIT_FAILURE);
				}
				break;

			case 'u': case 'i':
				fprintf (stderr, "tcpspy: -%c option is obsolete\n", ch);
				/* fall through to usage message */
			default:
				usage();
		}
	}

	argc -= optind;
	argv += optind;

	/*
	 * Become an unprivileged user for safety purposes if requested.
	 */
	if ((dropgroup == (uid_t) -1) && (defgroup != (uid_t) -1))
		dropgroup = defgroup;
	if (dropgroup != (gid_t) -1) {
		if (setgid (dropgroup) < 0) {
			fprintf (stderr, "tcpspy: setgid: %s\n", strerror (errno));
			exit (EXIT_FAILURE);
		}
	}
	if (dropuser != (uid_t) -1) {
		if (setuid (dropuser) < 0) {
			fprintf (stderr, "tcpspy: setuid: %s\n", strerror (errno));
			exit (EXIT_FAILURE);
		}
	}

	/*
	 * Become a daemon by double-forking and detaching completely from
	 * the terminal.
	 */

	if (debug == 0) {
		pid_t p;

		/* 1st fork */
		p = fork();
		if (p < 0) {
			fprintf (stderr, "tcpspy: fork: %s\n",
					strerror (errno));
			exit (EXIT_FAILURE);
		} else if (p != 0)
			exit (0);

		/* 2nd fork */
		p = fork();
		if (p < 0) {
			fprintf (stderr, "tcpspy: fork: %s\n",
					strerror (errno));
			exit (EXIT_FAILURE);
		} else if (p != 0) {
			fprintf (stderr, "tcpspy 1.7d started (pid %d)\n", 
					(int) p);
			exit (EXIT_SUCCESS);
		}

		ioctl (STDIN_FILENO, TIOCNOTTY, NULL);
		close (STDIN_FILENO); 
		close (STDOUT_FILENO); 
		close (STDERR_FILENO);
		setpgid (0, 0);
		chdir ("/");
	} else
		fprintf (stderr, "tcpspy 1.7d started (debug)\n");

	signal (SIGTERM, stopsig);
	if (debug != 0) signal (SIGINT, stopsig);

	/*
	 * Initialisation's done, start watching for connections.
	 */

	openlog ("tcpspy", LOG_PID, facility);
	if (debug == 0)
		/*
		 * .. for the benefit of those reading the logs. Useless to
		 * people running it in debug mode.
		 */
		logmsg ("tcpspy started, monitoring connections");

	if (ct_init (&old) == 0) panic ("ct_init failed");
	if (ct_read (&old) == 0) panic ("ct_read failed");
	
	while (stopped == 0) {
		struct timeval tv1, tv2;
		static double elapsed = 0.0;
		static int slow_warn = 0;

		gettimeofday (&tv1, NULL);

		if (ct_init (&new) == 0) panic ("ct_init failed");
		if (ct_read (&new) == 0) panic ("ct_read failed");
		compare (&old, &new);
		if (ct_free (&old) == 0) panic ("ct_free failed");
		memcpy (&old, &new, sizeof (conntable_t));

		gettimeofday (&tv2, NULL);

		/*
		 * If the time taken to poll the currently open connections is longer than
		 * the time between checks, emit a warning message.
		 */
		elapsed += (double) ((tv2.tv_sec - tv1.tv_sec) * 1000000 + (tv2.tv_usec - tv1.tv_usec));
		elapsed /= 2;
		if ((elapsed > (double)(interval * 1000)) && (slow_warn == 0)) {
			slow_warn = 1;
			logmsg ("warning: running slowly, suggest increasing interval with -I option");
		}

		usleep (interval * 1000);
	}

	if (ct_free (&old) == 0) panic ("ct_free failed");
	closelog ();

	return EXIT_SUCCESS;
}
